internal/impl/mail/input_pop3.go

```diff
@@ -0,0 +1,300 @@
+			return newPop3(conf, mgr.Logger())
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type pop struct {
+	client       *pop3.Client
+	username     string
+	password     string
+	count        int
+	interval     time.Duration
+	maxBodySize  int
+	listPageSize int
+	messageChan  chan *envelope
+	done         chan struct{}
+	ack          chan int
+	log          *service.Logger
+}
+
+type envelope struct {
+	message *service.Message
+	ID      int
+}
+
+// newPop creates a new pop3 input type.
+func newPop3(conf *service.ParsedConfig, log *service.Logger) (*pop, error) {
+	const errorPattern = "error retrieving %s: %v"
+	host, err := conf.FieldString(hostParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, hostParam, err)
+	}
+	port, err := conf.FieldInt(portParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, portParam, err)
+	}
+	username, err := conf.FieldString(usernameParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, usernameParam, err)
+	}
+	password, err := conf.FieldString(passwordParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, passwordParam, err)
+	}
+	tlsEnabled, err := conf.Namespace(tlsParam).FieldBool(tlsEnabledParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, tlsEnabledParam, err)
+	}
+	tlsSkipCertVerify, err := conf.Namespace(tlsParam).FieldBool(tlsSkipCertVerifyParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, tlsSkipCertVerifyParam, err)
+	}
+	count, err := conf.Namespace(pullParam).FieldInt(countParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, countParam, err)
+	}
+	interval, err := conf.Namespace(pullParam).FieldDuration(intervalParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, intervalParam, err)
+	}
+	maxInFlight, err := conf.FieldInt(maxInFlightParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, maxInFlightParam, err)
+	}
+	timeout, err := conf.FieldDuration(timeoutParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, timeoutParam, err)
+	}
+	maxBodySize, err := conf.FieldInt(maxBodySizeParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, maxBodySizeParam, err)
+	}
+	listPageSize, err := conf.FieldInt(listPageSizeParam)
+	if err != nil {
+		return nil, fmt.Errorf(errorPattern, listPageSizeParam, err)
+	}
+
+	client := pop3.New(pop3.Opt{
+		Host:          host,
+		Port:          port,
+		TLSEnabled:    tlsEnabled,
+		TLSSkipVerify: tlsSkipCertVerify,
+		DialTimeout:   timeout,
+	})
+
+	p := &pop{
+		client:       client,
+		username:     username,
+		password:     password,
+		maxBodySize:  maxBodySize,
+		listPageSize: listPageSize,
+		count:        count,
+		interval:     interval,
+		messageChan:  make(chan *envelope, maxInFlight),
+		done:         make(chan struct{}),
+		ack:          make(chan int, maxInFlight),
+		log:          log,
+	}
+
+	return p, nil
+}
+
+var onlyOnce sync.Once
```

Comment:
Looks like this needs to be owned by an individual `*pop`, otherwise multiple instances of this input would be ineffective.

Reply:
Fixed.